keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1330 @@
+since its workloads support surging to another node, which is the safest way to migrate. This would
+help to eliminate downtime not only for single replica applications, but for HA applications as
+well.
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without
+any required manual interventions. I want to have an ability to manually switch between the
+maintenance phases (Planning, Cordon, Drain, Drain Complete, Maintenance Complete). I also want to
+observe the node drain via the API and check on its progress. I also want to be able to discover
+workloads that are blocking the node drain.
+
+#### Story 2
+
+As an application owner, I want to run single replica applications without disruptions and have the
+ability to easily migrate the workload pods from one node to another.
+
+#### Story 3
+
+Cluster or node autoscalers that take on the role of `kubectl drain` want to signal the intent to
+drain a node using the same API and provide a similar experience to the CLI counterpart.
+
+#### Story 4
+
+I want to be able to use a similar approach for general descheduling of pods that happens outside
+of node maintenance.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+
+#### Cancelled Node Maintenance and Impact on Applications
+
+When node maintenance is cancelled (reaches `Complete` stage without all of its pods terminating),
+the node maintenance controller will remove `EvacuationRequested` conditions on the targeted pods.
+
+1. Pods that have not yet initiated evacuation (no `EvacuationInitiated` condition) will
+   continue to run unchanged.
+2. Pods that have initiated evacuation have an option for their controller to either cancel the
+   evacuation and remove the `EvacuationInitiated` condition, or keep the condition and finish the
+   evacuation.
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Kubectl
+
+`kubectl drain` command will be changed to create a NodeMaintenance object instead of marking the
+node unschedulable. We will also change the implementation to skip applications that support
+workload migration. This will be detected by observing a `EvacuationRequested` condition on the pod
+and the subsequent appearance of `EvacuationInitiated` condition within a reasonable timeframe (3m).
```

Comment:
How the 3m timeframe was selected?

Reply:
Additionally, is there any way to avoid this wait-and-fallback behaviour for everything else, e.g. StatefulSets? Adding ~3m delay to drains/maintenance would be frustrating. Perhaps a way of a controller opting-in to say that it supports this evacuation requested api or not, and immediately failing back to previous logic if not?

Reply:
Nevermind - it seems that the design has progressed significantly in https://github.com/kubernetes/enhancements/pull/4565#issuecomment-2045722703 towards dedicated resources with evacuators now advertising which pods they are able to evacuate.

Reply:
yup, there is no initial timeout anymore and the evacuation should be as responsive as the eviction